### ISSUE TITLE
belongsTo relationship causes "used previously in the same computation" error

### ISSUE THREAD

**Issue by wongpeiyi** [emberjs/ember.js#19003]
First posted in https://github.com/emberjs/data/issues/7196 but was referred here

When an API payload is received with an association not already set on a Ember Data model, the following error is seen:

```
Error: Assertion Failed: You attempted to update `content` on `<(unknown):ember173>`, 
but it had already been used previously in the same computation.  Attempting to update 
a value after using it in a computation can cause logical errors, infinite 
revalidation bugs, and performance issues, and is not supported.

`content` was first used:

- While rendering:
  
  application
    foo
      this.dataLength

Stack trace for the update:
    at dirtyTagFor (validator.js:514)
    at markObjectAsDirty (index.js:637)
    at notifyPropertyChange (index.js:675)
    at set (index.js:1622)
    at Proxy.set (observable.js:176)
    at InternalModel._updatePromiseProxyFor (-private.js:4379)
    at InternalModel.getBelongsTo (-private.js:4272)
```

This happens when the model is being consumed in a component template (via a getter) that depends on `post.user`

### Reproduction

https://github.com/wongpeiyi/data-computation

### Description

- When creating and saving a Post, no User association is set yet
- The API (mirage) sets a User and responds with a payload
- Ember Data updates the Post model
- Because the model is being consumed by the component getter that depends on `filterBy('user.id')`, the error is thrown

The error is not thrown when:
- Not filtering (or not consuming the user association)
- Using `{{@data.length}}` in the template instead of the component getter
- Using a getter on the controller without a component

This seems like a use case that should "just work"

### Versions

```cli
yarn list v1.19.2
warning Filtering by arguments is deprecated. Please use the pattern option instead.
└─ ember-source@3.18.1
✨  Done in 0.72s.
yarn list v1.19.2
warning Filtering by arguments is deprecated. Please use the pattern option instead.
└─ ember-cli@3.18.0
✨  Done in 0.65s.
yarn list v1.19.2
├─ @ember-data/adapter@3.18.0
├─ @ember-data/canary-features@3.18.0
├─ @ember-data/debug@3.18.0
├─ @ember-data/model@3.18.0
├─ @ember-data/private-build-infra@3.18.0
├─ @ember-data/record-data@3.18.0
├─ @ember-data/rfc395-data@0.0.4
├─ @ember-data/serializer@3.18.0
├─ @ember-data/store@3.18.0
├─ babel-plugin-ember-data-packages-polyfill@0.1.2
└─ ember-data@3.18.0
✨  Done in 0.63s.
```

**Comment by snewcomer** [comment]
I think this is a really interesting conversation, especially as we all rely more and more on autotracking inter-opting with things that implement [Observables](https://api.emberjs.com/ember/3.18/classes/EmberArray/methods/filterBy?anchor=filterBy).

My initial feeling was that b/c we are mixing paradigms of an array that implements `Observable` (push) with autotracking (pull), a solution shown [here](https://github.com/emberjs/data/issues/7196#issuecomment-638570207) might be necessary but not preferred.

@pzuraq Is this true that some solutions might not work with autotracking and have to be re-thought in app code or do we have preferred mechanism of fixing the source as detailed towards the bottom of this [comment](https://github.com/emberjs/data/issues/7196#issuecomment-638671313) without a full re-write towards autotracking semantics?  Perhaps a more general question as well is - are Observable friendly Array-like behaviours necessarily at odds with autotracking or is it a case by case basis like this one?

**Comment by snewcomer** [comment]
https://github.com/emberjs/ember.js/pull/19009/commits/aa607a6d9e4d6b1cc3d2bb69fac2bcf102e546c8

This commit might be closer to what we are looking for.  Lastly, need to trigger glimmer tracking/untrack.  I'll see what other steps we might need to take to get this to fail...

**Comment by runspired** [comment]
I've updated the investigation in the original ticket here: https://github.com/emberjs/data/issues/7196#issuecomment-646909813

TL;DR user error + mirage funkiness results in this.

However, there is also an issue in glimmer-vm in the DEBUG environment wherein having an unstable getter accessed multiple times by a template is problematic only in DEBUG and not in production where PropertyReferences are merged. See https://github.com/glimmerjs/glimmer-vm/pull/1111 for more details.

**Comment by snewcomer** [comment]
Thanks a lot for the explanation!  It was really useful.  

Finally got a test reproduction working as well!  Initially reproduced with a list of EmberData like models like the reproduction in this issue. However, pruned the test case to show the simplest path to hitting the error -   

https://github.com/emberjs/ember.js/pull/19009/files
